src/weighted_graph.py

```diff
@@ -0,0 +1,191 @@
+        g.del_node(n): deletes the node 'n' from the graph, raises an error if no such node exists
+        g.del_edge(n1, n2): deletes the edge connecting 'n1' and 'n2' from the graph, raises an error if no such edge exists
+        g.has_node(n): True if node 'n' is contained in the graph, False if not.
+        g.neighbors(n): returns the list of all nodes connected to 'n' by edges, raises an error if n is not in g
+        g.adjacent(n1, n2): returns True if there is an edge connecting n1 and n2, False if not, raises an error if either of the supplied nodes are not in g
+        g.depth_first_traversal(start): Returns the path list for the entire graph with a depth first traversal.
+        g.breadth_first_travers(start): Returns the path list for the entire graph with a breadth first traversal.
+
+    """
+
+    def __init__(self):
+        """Initialize a weighted graph data structure."""
+        self._gdict = {}
+
+    def nodes(self):
+        """Return a list of all nodes in the weighted graph."""
+        return list(self._gdict)
+
+    def edges(self):
+        """Return a list of all edges in the weighted graph."""
+        edge_list = []
+        for key in self._gdict:
+            for neighbor in list(self._gdict[key]):
+                edge_list.append((key, neighbor, self._gdict[key][neighbor]))
+        return edge_list
+
+    def add_node(self, n):
+        """Add a new node to the weighted graph."""
+        self._gdict.setdefault(n, OrderedDict())
+
+    def add_edge(self, n1, n2, weight=1):
+        """Add a new edge to the weighted graph connecting n1 and n2."""
+        if n1 == n2:
+            raise KeyError("Edges cannot connect a node to itself.")
+        if n1 in self._gdict and n2 in self._gdict[n1]:
+            return
+        if type(weight) is not int:
+            raise TypeError("weight must be an integer")
+        self._gdict.setdefault(n1, OrderedDict())[n2] = weight
+        self.add_node(n2)
+
+    def del_node(self, n):
+        """Delete node n from the weighted graph."""
+        try:
+            del self._gdict[n]
+            for key in self._gdict:
+                if n in self._gdict[key]:
+                    del self._gdict[key][n]
+        except KeyError:
+            raise KeyError("Can't delete a node if it doesn't already exist.")
+
+    def del_edge(self, n1, n2):
+        """Delete the edge connecting node 1 and node 2."""
+        try:
+            del self._gdict[n1][n2]
+        except KeyError:
+            raise KeyError("That edge does not exist.")
+
+    def has_node(self, n):
+        """Return True if node n exists in the weighted graph."""
+        if n in self._gdict:
+            return True
+        return False
+
+    def neighbors(self, n):
+        """Return a list of all neighbors of node n."""
+        try:
+            return list(self._gdict[n])
+        except KeyError:
+            raise KeyError("Node {} does not exist".format(n))
```

Comment:
Even though you are using keys under the hood, I think a ValueError would be more appropriate here.